providers/nextcloud/nextcloud.go

```diff
@@ -0,0 +1,211 @@
+	return NewCustomisedURL(
+		clientKey,
+		secret,
+		callbackURL,
+		nextcloudURL+"/apps/oauth2/authorize",
+		nextcloudURL+"/apps/oauth2/api/v1/token",
+		nextcloudURL+"/ocs/v2.php/cloud/user?format=json",
+		scopes...,
+	)
+}
+
+// Name is the name used to retrieve this provider later.
+func (p *Provider) Name() string {
+	return p.providerName
+}
+
+// SetName is to update the name of the provider (needed in case of multiple providers of 1 type)
+func (p *Provider) SetName(name string) {
+	p.providerName = name
+}
+
+func (p *Provider) Client() *http.Client {
+	return goth.HTTPClientWithFallBack(p.HTTPClient)
+}
+
+// Debug is a no-op for the nextcloud package.
+func (p *Provider) Debug(debug bool) {}
+
+// BeginAuth asks Gitlab for an authentication end-point.
+func (p *Provider) BeginAuth(state string) (goth.Session, error) {
+	return &Session{
+		AuthURL: p.config.AuthCodeURL(state),
+	}, nil
+}
+
+// FetchUser will go to Gitlab and access basic information about the user.
+func (p *Provider) FetchUser(session goth.Session) (goth.User, error) {
+	sess := session.(*Session)
+	user := goth.User{
+		AccessToken:  sess.AccessToken,
+		Provider:     p.Name(),
+		RefreshToken: sess.RefreshToken,
+		ExpiresAt:    sess.ExpiresAt,
+	}
+
+	if user.AccessToken == "" {
+		// data is not yet retrieved since accessToken is still empty
+		return user, fmt.Errorf("%s cannot get user information without accessToken", p.providerName)
+	}
+
+	req, err := http.NewRequest("GET", p.profileURL, nil)
+	if err != nil {
+		fmt.Println(err)
+		return user, err
+	}
+	req.Header.Set("Authorization", "Bearer "+sess.AccessToken)
+	response, err := p.Client().Do(req)
+
+	if err != nil {
+		if response != nil {
+			response.Body.Close()
+		}
+		fmt.Println(err)
+		return user, err
+	}
+
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return user, fmt.Errorf("%s responded with a %d trying to fetch user information", p.providerName, response.StatusCode)
+	}
+
+	bits, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		fmt.Println(err)
+		return user, err
+	}
+
+	err = json.NewDecoder(bytes.NewReader(bits)).Decode(&user.RawData)
+	if err != nil {
+		fmt.Println(err)
+		return user, err
+	}
+
+	err = userFromReader(bytes.NewReader(bits), &user)
+
+	fmt.Println(err)
+	return user, err
+}
+
+func newConfig(provider *Provider, authURL, tokenURL string, scopes []string) *oauth2.Config {
+	c := &oauth2.Config{
+		ClientID:     provider.ClientKey,
+		ClientSecret: provider.Secret,
+		RedirectURL:  provider.CallbackURL,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  authURL,
+			TokenURL: tokenURL,
+		},
+		Scopes: []string{},
+	}
+
+	if len(scopes) > 0 {
+		for _, scope := range scopes {
+			c.Scopes = append(c.Scopes, scope)
+		}
+	}
+	return c
+}
+
+func userFromReader(r io.Reader, user *goth.User) error {
+	u := struct {
+		Ocs struct {
+			Data struct {
+				EMail       string `json:"email"`
+				DisplayName string `json:"display-name"`
```

Comment:
Does NextCloud really return the the JSON payload with the key `"display-name"` (not `"display_name"`)? That's a bit odd to me, so I'm just wondering if this is a typo or if they really do use that key.

Reply:
Yes, they really do. The API is a bit tricky to find in their code, but [this test does expect it](https://github.com/nextcloud/server/blob/42121fe4d71237f769831e2b33def779e0566e7c/apps/provisioning_api/tests/Controller/UsersControllerTest.php#L3056).